src/platform/plugins/shared/discover/public/application/main/state_management/tabs_storage_manager.ts

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import {
+  createStateContainer,
+  type IKbnUrlStateStorage,
+  type ReduxLikeStateContainer,
+  syncState,
+} from '@kbn/kibana-utils-plugin/public';
+import type { TabItem } from '@kbn/unified-tabs';
+import type { Storage } from '@kbn/kibana-utils-plugin/public';
+import { TABS_STATE_URL_KEY } from '../../../../common/constants';
+import type { TabState } from './redux/types';
+import { createTabItem } from './redux/utils';
+
+const TABS_LOCAL_STORAGE_KEY = 'discover.tabs';
+
+interface TabStateInLocalStorage {
+  tabState: Pick<TabState, 'id' | 'label' | 'dataViewId' | 'appState' | 'globalState'>;
+  version: string;
+}
+
+interface TabsStateInLocalStorage {
+  openTabs: TabStateInLocalStorage[];
+  closedTabs: Array<TabStateInLocalStorage & { closedAt: number }>;
+}
+
+interface TabsInternalStatePayload {
+  allTabs: TabState[];
+  selectedTabId: string;
+}
+
+export interface TabsStorageState {
+  id?: string; // syncing the selected tab id with the URL
+}
+
+export interface TabsStorageManager {
+  /**
+   * Supports two-way sync of the selected tab id with the URL.
+   * Currently, we use it only one way - from internal state to URL.
+   */
+  urlStateContainer: ReduxLikeStateContainer<TabsStorageState>;
+  startUrlSync: () => () => void;
+  persistLocally: (props: TabsInternalStatePayload) => Promise<void>;
+  updateTabStateLocally: (
+    tabId: string,
+    tabState: Pick<TabState, 'globalState' | 'appState'>
+  ) => void;
+  loadLocally: (props: {
+    defaultTabState: Omit<TabState, keyof TabItem>;
+  }) => TabsInternalStatePayload;
+}
+
+export const getTabsStorageManager = ({
+  urlStateStorage,
+  storage,
+  onChanged,
+}: {
+  urlStateStorage: IKbnUrlStateStorage;
+  storage: Storage;
+  onChanged?: (nextState: TabsStorageState) => void; // can be called when selectedTabId changes in URL to trigger app state change if needed
+}): TabsStorageManager => {
+  const urlStateContainer = createStateContainer<TabsStorageState>({});
```

Comment:
I partly feel like we should be migrating away from these custom state containers entirely and relying more on the Redux store, but I'm not really sure on the right approach. We might be able to sync from Redux directly to `urlStateStorage` using a `tabId` query param or similar, but maybe that has its own drawbacks. Just seems like we're introducing another source of truth with `TabsStorageState`.